File: exptr_api/internal/storage/postgres/categories.go
package postgres

import (
	"errors"
	"fmt"

	"alex_gorbunov_exptr_api/internal/domain"
	"alex_gorbunov_exptr_api/internal/models"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

func (s *Storage) CreateCategory(category *models.CategoryRequest) error {
	const fn = "storage.postgresql.CreateCategory"

	cat := domain.Category{
		UserID: uuid.MustParse(category.UserID),
		Name:   category.Name,
		Type:   category.Type,
		Color:  category.Color,
		Icon:   category.Icon,
	}

	result := s.db.Create(&cat)
	if result.Error != nil {
		return fmt.Errorf("%s: %w", fn, result.Error)
	}

	return nil
}

func (s *Storage) UpdateCategory(category *domain.Category) error {
	const fn = "storage.postgresql.UpdateCategory"

	result := s.db.Model(&domain.Category{}).Where("id = ?", category.ID).Updates(map[string]interface{}{
		"user_id": category.UserID,
		"name":    category.Name,
		"type":    category.Type,
		"color":   category.Color,
		"icon":    category.Icon,
	})

	if result.Error != nil {
		return fmt.Errorf("%s: %w", fn, result.Error)
	}

	if result.RowsAffected == 0 {
		return fmt.Errorf("%s: category not found", fn)
	}

	return nil
}

func (s *Storage) GetCategories(userID uuid.UUID) ([]domain.Category, error) {
	const fn = "storage.postgresql.GetCategories"

	var categories []domain.Category
	result := s.db.Where("user_id = ?", userID).Find(&categories)
	if result.Error != nil {
		return nil, fmt.Errorf("%s: %w", fn, result.Error)
	}

	return categories, nil
}

func (s *Storage) GetCategoryByID(id uuid.UUID) (*domain.Category, error) {
	const fn = "storage.postgresql.GetCategoryByID"

	var category domain.Category
	result := s.db.Where("id = ?", id).First(&category)
	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return nil, fmt.Errorf("%s: category not found", fn)
		}
		return nil, fmt.Errorf("%s: %w", fn, result.Error)
	}

	return &category, nil
}

func (s *Storage) DeleteCategory(id uuid.UUID) error {
	const fn = "storage.postgresql.DeleteCategory"

	// First check if category exists
	var category domain.Category
	result := s.db.Where("id = ?", id).First(&category)
	if result.Error != nil {
		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
			return fmt.Errorf("%s: category not found", fn)
		}
		return fmt.Errorf("%s: %w", fn, result.Error)
	}

	// Delete the category (soft delete due to gorm.DeletedAt in BaseEntity)
	result = s.db.Delete(&category)
	if result.Error != nil {
		return fmt.Errorf("%s: %w", fn, result.Error)
	}

	return nil
}

File: exptr_api_p/exptr_api/repositories/categories.py
from sqlalchemy.orm import Session

from exptr_api.entities import categories
from exptr_api.repositories.database import categories as db_categories

class CategoriesRepository:
  def __init__(self, session: Session):
    self.session = session

  def create_category(self, category: categories.CategoryRequest):
    db_category = db_categories.Category(
      user_id=category.user_id,
      name=category.name,
      type=category.type,
      created_at=category.created_at,
      updated_at=category.updated_at,
      color=category.color,
      icon=category.icon
    )
    self.session.add(db_category)
    self.session.commit()
    self.session.refresh(db_category)
    return db_category

  def update_category(self, category: categories.Category):
    db_category = self.session.query(db_categories.Category).filter(db_categories.Category.id == category.id).first()
    db_category.user_id = category.user_id
    db_category.name = category.name
    db_category.type = category.type
    db_category.updated_at = category.updated_at
    db_category.color = category.color
    db_category.icon = category.icon
    self.session.commit()
    self.session.refresh(db_category)
    return db_category

  def get_categories(self, user_id: int):
    return self.session.query(db_categories.Category).filter(db_categories.Category.user_id == user_id)

  def delete_category(self, category_id: int):
    self.session.query(db_categories.Category).filter(db_categories.Category.id == category_id).delete()
    self.session.commit()